### ISSUE TITLE
Create temporary resources during terraform apply (automatically destroy it at the end of apply)

### ISSUE THREAD

**Issue by ejbp** [hashicorp/terraform#32067]
### Terraform Version

```shell
Terraform v1.3.1
```

### Use Cases

I want to be able to run commands on an ec2 instance (for example) on the pipeline. Since the pipeline runner is not inside my vpc, and the ec2 instance should to not be open to outside, I need to temporary create a security group, during the apply, that will allow ssh inbound from outside in order to be able to run ansible commands on that ec2 instance (for example) during the apply. At the end, I would like that terraform automatically removes that security group for me.

### Attempted Solutions

Currently I have the destroy command on that target, on the next job after the apply. I don't like this solution for many reasons, but specially because if you run terraform apply on your local computer and the developer forgets to destroy the security group manually, the ec2 will be opened to outside until someone runs pipeline or removes it manually.    

### Proposal

add a generic option for every resource (that makes sense) where you can add "destroy_me_after_apply = true". The default would be false, of course. If this option is settled to true, then the resource will be created during the apply and will be removed at the end of the apply.

**Comment by crw** [comment]
Thanks for this feature request!

**Comment by bartodes** [comment]
Hello! There is a way to get this done
Do you still need help?

**Comment by ejbp** [comment]
Hi. Yes, please....

**Comment by bartodes** [comment]
# Context
This was done in Azure, but it's going to work for any Cloud Provider.

I needed to make a module of a postgresql flexibe server, that had an 'auto import' feature (this was to import the dump as soon as the resource) So because the database server hadn't a public-ip, i had to deploy 3 resources and then destroy them but once the import was done and I wanted to do a <terraform apply> to deploy something else, the resources from the auto_import feature would begin to create again.

For the explanation lets just have a Resource Group, a Virtual Machine. So we want both to be deleted at some point in the apply, but we don't want to make an adjusment, then re-apply and have them deployed one more time. 

The null_resource one is for deleting the other 2 resources automatically and we need it as "proof" that the other resources have been created once.

```tf
module "import_rg" {
  source = "../../resource-groups"

  count = var.auto_import.enabled ? !local.done ? 1 : 0 : 0

  name     = "import-rg"
  location = var.location
}

module "import_vm" {
  source = "../../compute/vm"

  count = var.auto_import.enabled ? !local.done ? 1 : 0 : 0

  [ ... ]
}

resource "null_resource" "database_import" {

  count = var.auto_import.enabled ? 1 : 0

  provisioner "local-exec" {
    command = trimspace(<<EOF
      az group delete --resource-group ${module.import_rg[0].name} -y -f Microsoft.Compute/virtualMachines
    EOF
    )
  }

  lifecycle {
    ignore_changes = all
  }
}
```

_The conditional in the count was intended so the auto_import feature could be enabled or desabled, if the argument was set to 'true' or 'false'. This way count will be only 1, when the value of 'var.auto_import.enabled' is set to 'true'. By default is set to false. The second condition in the count of the Resource Group and the VM, will be explained in the solution section._

# Solution

The solution is simple. For terraform not to create this resources again in the apply, you will need to read the tfstate file with a data source, to analize if the resource was or was not created with out having a clasic *Cycle Error*. To get this we will need 3 more things. An output, a data source and a locals block. 

The output is for modifying the tfstate file, the data source is for reading it and the locals is to add a condition for the final value. 

We declare an output, this one has to be declared in the root .tf files. The ones you use to execute all tf
#### output.tf
```tf
output "done" {
  value = null_resource.database_import != [] ? null_resource.database_import[0].id != null ? true : false : false
}
```
#### data.tf
```tf
data "terraform_remote_state" "check" {
  backend = "<name_of_backend>"
  config = {...}
}
```
#### locals.tf
```tf
locals {
  done = can(data.terraform_remote_state.check.outputs["done"]) ? data.terraform_remote_state.check.outputs.done : false
}
``` 
Alright, now we have "local.done" which in case that the output does exists in the tfstate file it will have its value, and if the output does not exists in the tfstate, local.done will equal to false.

With this finally we can evaluate local.done value and not have the resources deployed once again!

# Results

The apply will look like this (*You will need to have the output declared in the root module, for all of this to work!*)

![Screenshot from 2023-10-24 18-21-03](https://github.com/hashicorp/terraform/assets/69482682/fb0832ae-0f55-4d56-93f6-bacaa40c4d4f)

*PD: I hope that with this example, you've got an idea of how to implement this for your case! Happy to help any time ;) And sorry for my english, is not my first language. Take care!*

**Comment by ejbp** [comment]
I think you misunderstood my question and probably it's because I didn't explain it correctly.

What I want is that during each run of the pipeline with the "terraform apply":

1. Create Resource A (imagine a security group that allows ssh port for the runner ip)
2. Do the update on all the other resources
3. Destroy the Resource A (on this apply)

I have a solution (workaround to do it) but I was suggesting to add this feature to terraform.

```terraform
resource "aws_security_group" "allow_ssh_to_runner" {
  name   = "allow-ssh-public"
  vpc_id = local.vpc_id

  ingress = [
    {
      description      = "Allow Any Inbound to ssh"
      protocol         = "tcp"
      from_port        = 22
      to_port          = 22
      cidr_blocks      = [xxx.xx.xx.xx/32]
    }
  ]
  
  #This is my suggestion
  lifecycle {
    ephemeral = true # or create_and_destroy=true
  }
  
}
```

**Comment by g-kartik** [comment]
Yes, this feature shall prove useful. For instance I am spinning up ephemeral docker containers for the setup of the main infrastructure.

**Comment by netikras** [comment]
I'd like to add to this request: make it so `terraform apply` could **delete** particular resources. 

I have a terraform script that installs a number of helm charts. However, the helm chart configuration is not fine-grained enough, and after `helm install` I need to remove some of the k8s resources, otherwise alerting will destroy my mailbox. And I need those resources to remain destroyed (e.g. a PrometheusRule monitoring for kube-proxy, when I have Cilium CNI that doesn't use kube-proxy; however, I can't fine-tune prometheus-helm-chart to specify which PrometheusRule's to exclude)

**Comment by bartodes** [comment]
> I think you misunderstood my question and probably it's because I didn't explain it correctly.
> 
> What I want is that during each run of the pipeline with the "terraform apply":
> 
> 1. Create Resource A (imagine a security group that allows ssh port for the runner ip)
> 2. Do the update on all the other resources
> 3. Destroy the Resource A (on this apply)
> 
> I have a solution (workaround to do it) but I was suggesting to add this feature to terraform.
> 
> ```hcl
> resource "aws_security_group" "allow_ssh_to_runner" {
>   name   = "allow-ssh-public"
>   vpc_id = local.vpc_id
> 
>   ingress = [
>     {
>       description      = "Allow Any Inbound to ssh"
>       protocol         = "tcp"
>       from_port        = 22
>       to_port          = 22
>       cidr_blocks      = [xxx.xx.xx.xx/32]
>     }
>   ]
>   
>   #This is my suggestion
>   lifecycle {
>     ephemeral = true # or create_and_destroy=true
>   }
>   
> }
> ```

Oh yes @ejbp, I misunderstood that you watend the workarround explained, my bad! This feature in lifecycle would be so nice to have.  Create a resource only one time, destroy the resource and it will never be deployed again. So cool 😎

I think they will have to develop some twisted logic in order to implement this feature. This is because the resource we want to set to `ephemeral = true` will have to destroy all infrastructure deployed by the resource itself (only the infrastructure mapped to the resource), but at the same time the resource will have to remain in the `created` state and not `deleted` once the `terraform apply` ends.

**Comment by ejbp** [comment]
@bartodes That would be a great feature too :) Sometimes you want to destroy a bunch of resources without destroying others and gives some work to accomplish that.

But the **ephemeral feature** would be a resource that only exists during the pipeline (ex: during the terraform apply lifecycle) but after that it would not exist anymore. I will try with more simple explanation:

1. The resource doesn't exist
2. run the terraform apply
3. during terraform apply the resource is created
4. when the terraform apply terminates, with error or not, the resource is automatically destroyed by the terraform
5. Now the resource doesn't exists again

if you run the terraform apply the previous points will repeat and it will happen again. The resource is created and destroyed during the lifecycle, even if it fails. There are many use cases where this feature is useful.

I hope now you can understand it.

**Comment by panfiva** [comment]
The solution should be flexible and provide support for temporary configurations, not just temporary resources.  So we need to support multi-state lifecycle that includes either update or delete in the second stage.

For example, when creating an AWS EC2 resource that needs to have direct Internet connectivity during provisioning, but go through a proxy after provisioned, we need to be able to perform the following:
- create a route table RT
- Stage 1: add default route to RT pointing to AWS Internet Gateway
- provision EC2
- provision proxy server
- Stage 2: update default route table to use proxy server network interface as default route

**Comment by AurimasNav** [comment]
I would use it for creating cloudflare DNS records with proxy = false, to be able to validate DNS ownership for other resources, and after other resources successfully deployed, destroy this cloudflare DNS record and re-create another with proxy enabled.

**Comment by pkuczynski** [comment]
There is `ephemeral` resource since v.1.10, however it not always works, like for example https://github.com/hashicorp/terraform-provider-azurerm/issues/29064.

**Comment by pkuczynski** [comment]
@crw where does `azurem` provider stands on your roadmap?

**Comment by crw** [comment]
AzureRM is owned by the Azure Provider team at HashiCorp and has its own roadmap. Their repository is, as previously linked, https://github.com/hashicorp/terraform-provider-azurerm/. I can't speak for their roadmap or timelines.

**Comment by pkuczynski** [comment]
Understood. I posted a feature request on their repo earlier on https://github.com/hashicorp/terraform-provider-azurerm/issues/29064. Let's see if they will react to it... Thank you @crw for promptly replies...
